odl/solvers/scalar/steplen.py

```diff
@@ -148,6 +148,21 @@ def __call__(self, x, direction, dir_derivative):
             The computed step length
         """
         alpha = 1.0
+
+        # Since all comparisons with nan returns False, in order to not get a
+        # problem with nan-values we throw an error.
+        if self.function(x) == float('nan'):
+            raise NameError('ObjFunIsNanError')
+
+        # Next step is done in order to get a step length that does not give
+        # cost nan or infinite. Note that for convex functions, if the current
+        # point and the current point + alpha * direction does not give nan or
+        # inf this guarantees that any point along the line combinding these
+        # will not be inf either.
+        while (self.function(x + alpha * direction) == float('nan') or
+                self.function(x + alpha * direction) == float('inf')):
```

Comment:
Didn't see this the first time: This is also very sub-optimal. It will almost always evaluate twice, even if both conditions are false. And the value is not reused in the next loop.

The value `fxstep = self.function(x + alpha * direction)` should be stored, then used in the condition here, and finally used as first value "proper" line search:
```python
if fxstep < fx + alpha * dir_derivative * self.discount:
    return alpha
else:
    num_iter = 1
while ...
```